File: frontend/src/lib/command-intent.ts
/** Mirror backend command_intent.can_auto_approve for UI auto-run gating. */

const FIX_MARKERS = [
  "systemctl enable",
  "systemctl disable",
  "systemctl restart",
  "systemctl start",
  "systemctl stop",
  "mount -o remount",
  "sed -i",
  "chmod ",
  "chown ",
  "chgrp ",
  "setfacl",
  "useradd",
  "groupadd",
  "apt-get clean",
  "journalctl --vacuum",
  "-exec chown",
  "-exec chmod",
];

function intentFromCommand(commandText: string): "diagnostic" | "fix" | "validate" {
  const t = (commandText || "").trim().toLowerCase();
  if (!t) return "diagnostic";
  if (t.includes("public-test") || (t.startsWith("curl ") && t.includes("health"))) {
    return "validate";
  }
  if (
    t.includes("chmod") ||
    t.includes("chown") ||
    t.includes("systemctl restart") ||
    t.includes("systemctl start") ||
    t.includes("systemctl enable") ||
    t.includes("sed -i")
  ) {
    return "fix";
  }
  return "diagnostic";
}

export function canAutoApprove(commandText: string, safetyStatus?: string): boolean {
  if (safetyStatus === "Blocked") return false;
  const text = (commandText || "").trim();
  if (!text) return false;
  const lower = text.toLowerCase();
  if (FIX_MARKERS.some((m) => lower.includes(m))) return false;
  return intentFromCommand(text) === "diagnostic";
}

export function commandIntentLabel(commandText: string): string {
  return intentFromCommand(commandText);
}

File: backend/app/agent/command_intent.py
"""Classify proposed commands — prefer explicit plan intent over substring heuristics."""
from __future__ import annotations

from app.agent.command_validator import is_valid_shell_command

_FIX_MARKERS = (
    "chmod",
    "chown",
    "systemctl restart",
    "systemctl start",
    "systemctl enable",
    "mount -o remount",
    "sed -i",
    "setfacl",
    "useradd",
    "groupadd",
)


def intent_from_command(command_text: str) -> str:
    """Return diagnostic | fix | validate for a shell command."""
    t = (command_text or "").strip().lower()
    if not t:
        return "diagnostic"
    if "public-test" in t or t.startswith("curl ") and "health" in t:
        return "validate"
    if any(m in t for m in _FIX_MARKERS):
        return "fix"
    return "diagnostic"


def is_fix_command(command_text: str, *, plan_intent: str | None = None) -> bool:
    if plan_intent:
        return plan_intent == "fix"
    return intent_from_command(command_text) == "fix"


def is_validate_command(command_text: str, *, plan_intent: str | None = None) -> bool:
    if plan_intent:
        return plan_intent == "validate"
    return intent_from_command(command_text) == "validate"


def is_valid_proposal_command(text: str) -> bool:
    return is_valid_shell_command(text)
